pkg/controller/attach_detach.go

```diff
@@ -0,0 +1,586 @@
+	if azVolumeAttachment == nil {
+		return nil
+	}
+
+	if azVolumeAttachment.Labels == nil {
+		azVolumeAttachment.Labels = map[string]string{}
+	}
+	azVolumeAttachment.Labels[azureutils.RoleLabel] = string(role)
+
+	if azVolumeAttachment.Status.Detail == nil {
+		return azVolumeAttachment
+	}
+	azVolumeAttachment.Status.Detail.Role = role
+
+	return azVolumeAttachment
+}
+
+func (r *ReconcileAttachDetach) updateStatusDetail(ctx context.Context, azVolumeAttachment *v1alpha1.AzVolumeAttachment, status map[string]string) *v1alpha1.AzVolumeAttachment {
+	if azVolumeAttachment == nil {
+		return nil
+	}
+
+	if azVolumeAttachment.Status.Detail == nil {
+		azVolumeAttachment.Status.Detail = &v1alpha1.AzVolumeAttachmentStatusDetail{}
+	}
+	azVolumeAttachment.Status.Detail.Role = azVolumeAttachment.Spec.RequestedRole
+	azVolumeAttachment.Status.Detail.PublishContext = status
+
+	return azVolumeAttachment
+}
+
+func (r *ReconcileAttachDetach) updateError(ctx context.Context, azVolumeAttachment *v1alpha1.AzVolumeAttachment, err error) *v1alpha1.AzVolumeAttachment {
+	if azVolumeAttachment == nil {
+		return nil
+	}
+
+	if err != nil {
+		azVolumeAttachmentError := &v1alpha1.AzError{
+			ErrorCode:    util.GetStringValueForErrorCode(status.Code(err)),
+			ErrorMessage: err.Error(),
+		}
+		if derr, ok := err.(*volerr.DanglingAttachError); ok {
+			azVolumeAttachmentError.ErrorCode = util.DanglingAttachErrorCode
+			azVolumeAttachmentError.CurrentNode = derr.CurrentNode
+			azVolumeAttachmentError.DevicePath = derr.DevicePath
+		}
+
+		azVolumeAttachment.Status.Error = azVolumeAttachmentError
+	}
+
+	return azVolumeAttachment
+}
+
+func (r *ReconcileAttachDetach) updateState(ctx context.Context, azVolumeAttachment *v1alpha1.AzVolumeAttachment, state v1alpha1.AzVolumeAttachmentAttachmentState) *v1alpha1.AzVolumeAttachment {
+	if azVolumeAttachment == nil {
+		return nil
+	}
+
+	if azVolumeAttachment.Status.State == state {
+		return azVolumeAttachment
+	}
+	azVolumeAttachment.Status.State = state
+	return azVolumeAttachment
+}
+
+func (r *ReconcileAttachDetach) updateWithRetry(ctx context.Context, attachmentName string, updateFunc func(*v1alpha1.AzVolumeAttachment)) error {
+	conditionFunc := func() (bool, error) {
+		azVolumeAttachment, err := azureutils.GetAzVolumeAttachment(ctx, r.client, r.azVolumeClient, attachmentName, r.namespace, true)
+		if err != nil {
+			klog.Errorf("failed to get AzVolumeAttachment (%s): %v", attachmentName, err)
+			// if failed to get azVolumeAttachment then escape the wait
+			return false, err
+		}
+
+		azVolumeAttachment = azVolumeAttachment.DeepCopy()
+		updateFunc(azVolumeAttachment)
+
+		if err := r.update(ctx, azVolumeAttachment); err != nil {
+			return false, nil
+		}
+		return true, nil
+	}
+	return wait.PollImmediateInfinite(updateAttemptInterval, conditionFunc)
```

Comment:
Wouldn't it better to set a timeout here? Or is there a reason for polling infinitely?

Reply:
I saw no intrinsic difference between polling infinitely vs. requeuing to reconciler infinitely. The problem I saw with the later approach was that sometimes, too many reconciler requests end up being enqueued for the same `AzVolumeAttachment` CRI and end up having majority of the workers run reconciliation that is destined to fail. 

But thinking twice, I see a bigger problem with the prior approach: it blocks the worker. So theoretically, our controller could have all its workers blocked with infinite `updateWithRetry`.

Reply:
If we want to save reconciliation cycles, can we use PollImmediate() instead? With that we could provide a fixed timeout so that the worker isn't blocked infinitely.

Reply:
Yes, I made the change in my local. I will push the commit along with other changes :)